README.md
# Nodepub v2.1.1

Create valid EPUB (v2) ebooks with metadata, contents and cover image.

## About Nodepub

Nodepub is a **Node** module which can be used to create **EPUB (v2)** documents.

* Files pass the [IDPF online validator](http://validator.idpf.org)
* Files meet Sigil's preflight checks
* Files open fine in iBooks, Adobe Digital Editions and Calibre
* Files open fine with the Kobo H20 ereader
* Files are fine as *KindleGen* input
* PNG cover images - recommend 600x800, 600x900, or similar as minimum
* Custom CSS can be provided
* Inline images within the EPUB
* Optionally generate your own contents page
* Front matter before the contents page
* Exclude sections from auto contents page and metadata-based navigation
* OEBPS and other 'expected' subfolders within the EPUB

## Installing Nodepub

Add it as with any other module:

``` sh
npm i nodepub
```

Import it for use:

``` javascript
var nodepub = require("nodepub");
```

## Using Nodepub

### Creating your content

Your document will consist of *metadata* and *sections*.
Metadata is an object with various properties detailing the book.
Sections are chunks of HTML that can be thought of as chapters.

Here's some sample metadata:

``` javascript
var metadata = {
  id: '278-123456789',
  title: 'Unnamed Document',
  series: 'My Series',
  sequence: 1,
  author: 'KA Cartlidge',
  fileAs: 'Cartlidge,KA',
  genre: 'Non-Fiction',
  tags: 'Sample,Example,Test',
  copyright: 'Anonymous, 1980',
  publisher: 'My Fake Publisher',
  published: '2000-12-31',
  language: 'en',
  description: 'A test book.',
  contents: 'Table of Contents',
  source: 'http://www.kcartlidge.com',
  images: ["../test/hat.png"]
};
```

* The `series` and `sequence` are not recognised by many readers (the *Calibre* properties are used).
* The `tags` become subjects in the final EPUB.
* For `published` note the year-month-day format.
* The `language` is the short *ISO* language name (`en`, `fr`, `de` etc).
* The `images` array is where you refer to all images used inside the book. Within your section HTML you always link to one all-containing flat folder: `<img src="../images/hat.png" />`.

### Populating EPUBs

Call `document` with a metadata object detailing your book plus the path of a cover image.

``` javascript
var epub = makepub.document(metadata, "./cover.png");
```

#### Choose to make your own contents page

```javascript
var makeContentsPage = function(links) {
  var contents = "<h1>Chapters</h1>";
  _.each(links, function (link) {
    if (link.itemType !== "contents") {
      contents += "<a href='" + link.link + "'>" + link.title + "</a><br />";
    }
  });
  return contents;
};
var epub = makepub.document(metadata, "./cover.png", makeContentsPage);
```

The `links` array which is passed to a callback consists of objects with the following properties:

* *title* - the title of the section to be linked to.
* *link* - the relative `href` within the EPUB.
* *itemType* - one of 3 types, those being *front* for front matter, *contents* for the contents page and *main* for the remaining sections.

The callback should return a string of HTML which will form the body of the contents page.

#### Optionally add some CSS

``` javascript
epub.addCSS(`p { text-indent: 0; } p+p { text-indent: 0.75em; }`);
```

#### Fill in the actual content

Call `addSection` with a title and HTML contents for each section (usually a chapter) plus options for whether to exclude it from the contents page list and/or to use it as front matter.

``` javascript
epub.addSection('Copyright', copyright, false, true);
epub.addSection('Chapter 1', "<h1>One</h1><p>...</p>");
```

Excluding from the contents page list allows you to add content which does not appear either in the auto-generated HTML table of contents or the metadata contents used directly by ereaders/software. This is handy for example when adding a page just after the cover containing the title and author - a common page which does not usually appear in the book contents.

### Creating a complete EPUB file ready for distribution

This is the simplest option.

``` javascript
epub.writeEPUB(
  function (e) { console.log("Error:", e); },
  '/folder', 'filename-without-extention',
  function () { console.log("No errors.") }
);
```

### Creating a folder containing all the files necessary to build the final EPUB yourself

If you need to do further work on your EPUB, this option writes the entire constituent files into a folder. When you write your EPUB out, it should consist of this folder's contents zipped up and renamed, with the sole proviso that the first file in the zip should be the `mimetype` file stored uncompressed.

``` javascript
epub.writeFilesForEPUB(
  '/folder-for-EPUB-files',
  (err) => { if (err) { console.log(err); } }
);
```

### Creating a Javascript object containing all the filenames and content needed for the final EPUB

Finally in a similar manner to the previous folder of files you can have the same set of content returned to you directly for you to do what you want with.

``` javascript
epub.getFilesForEPUB(
  function (err, epubFiles) {
    console.log("Files array:", epubFiles);
  }
);
```

---

## Substitutions

A simple form of text substitution is supported. At any point in your section's content you may include placeholders like `[[COPYRIGHT]]`. When the EPUB is generated any such placeholders which match the *capitalised* name of a *metadata* entry are replaced with the corresponding metadata value.

For example, you could have a "Thanks for Reading" page at the end of your book which has the following markup:

``` html
<p>Thanks for reading <strong>[[TITLE]]</strong> by <em>[[AUTHOR]]</em>.</p>
```

When the EPUB is produced and opened, if your metadata was set as per the book "The Hobbit" by "JRR Tolkien" it will look like:

---

Thanks for reading **The Hobbit** by *JRR Tolkien*.

---

This means you can re-use content across multiple books or refer to the author/title/series/whatever at any point within the book without worrying about consistency or maintenance.

## An example

In the top folder (containing the *package.json* file) run the following.

``` javascript
npm run example
```

This will generate various outputs and advise you of the location.

---

---

## For nodepub developers only

### Tests and example

In the top folder (containing the *package.json* file) run one of the following.

``` javascript
npm test
npm run example
```

### Code quality

In addition to the tests I use **ES Lint** and **Editor Config**. Make sure your editor/ide has support for the `.editorconfig` file, and when changing code do the following afterwards.

``` javascript
npm run lint
```

This will auto-fix what it can. You will have less issues if your editor/IDE also has an ES Lint integration.

*Visual Studio Code* has both of the above via plugins.

### Notes

* The tests mostly stub *fs* where used. However at one point they do actually write a final EPUB document. This means that (a) the test process needs write access to the test folder and (b) an actual file is generated.

* Whilst the *process* of generating EPUBs is tested, the *final EPUB* is not; I have manually tested it via the [IDPF Validator](http://validator.idpf.org/). The actual testing of an EPUB file is already sufficiently covered by the *epubcheck* tool which that site uses, and I have not added it as an integration test.

* You may find it helpful to look at the `example/example.js` file.

* For viewing generated metadata and content both Microsoft Edge and Calibre have good EPUB readers.

## Reminder

This is a utility module, not a user-facing one. In other words it is assumed that the caller has already validated the inputs. Only basic ommission checks are performed.


RELEASES.md
# Releases

## v2.1.1 - 2019/03/20

* Updated `package.json` to add `src` folder to `start` entry

## v2.1.0 - 2019/03/20

* Updated dependencies
* Restructured source folders
* Switched from Jasmine tests to Mocha
* Added *npm* scripts for running `eslint` and the example

### Breaking changes

* The `makeContentsPage` callback function will no longer receive the default markup as a second parameter.

## v2.0.7

* Thanks to [Harold Treen](https://github.com/haroldtreen) the API has switched to being asynchronous

## v1.0.7

* This version introduces stability at the expense of minor breaking changes, consisting almost entirely of renames from chapter to section but with some of the pre-generated pages of earlier versions eliminated. The result is more abstracted but also more flexible, whilst also retaining most of it's simplicity.


example/example.js
/* eslint-disable no-console */
var nodepub = require("../src/index.js"),
  _ = require("lodash");

// Metadata example.
var metadata = {
  id: '278-123456789',
  title: 'Unnamed Document',
  series: 'My Series',
  sequence: 1,
  author: 'Anonymous',
  fileAs: 'Anonymous',
  genre: 'Non-Fiction',
  tags: 'Sample,Example,Test',
  copyright: 'Anonymous, 1980',
  publisher: 'My Fake Publisher',
  published: '2000-12-31',
  language: 'en',
  description: 'A test book.',
  contents: 'Chapters',
  source: 'http://www.kcartlidge.com',
  images: ['test/hat.png']
};

var copyright = `<h1>[[TITLE]]</h1>
<h2>[[AUTHOR]]</h2>
<h3>&copy; [[COPYRIGHT]]</h3>
<p>
	All rights reserved.
</p>
<p>
	No part of this book may be reproduced in any form or by any electronic or
	mechanical means, including information storage and retrieval systems, without
	written permission from the author, except where covered by fair usage or
	equivalent.
</p>
<p>
	This book is a work of fiction.
	Any resemblance to actual persons	(living or dead) is entirely coincidental.
</p>
`;

var more = `<h1>More Books to Read</h1>
<h2>Thanks for reading <em>[[TITLE]]</em>.</h2>
<p>
	I hope you enjoyed the book, but however you felt please consider leaving a
	review where you purchased it and help other readers discover it.
</p>
<p>
	You can find links to more books (and often special offers/discounts) by
	visiting my site at <a href="http://kcartlidge.com/books">KCartlidge.com/books</a>.
</p>
`;

var about = `<h1>About the Author</h1>
<p>
	This is just some random blurb about the author.
</p>
<p>
	You can find more at the author's site.
</p>
<p>
	Oh, and here's a picture of a hat:
</p>
<p>
	<img src="../images/hat.png" alt="A hat." />
</p>
`;

// Dummy text (lorem ipsum).
var lipsum = "<p>Lorem ipsum dolor sit amet, consectetur adipiscing elit. Suspendisse mattis iaculis pharetra. Proin malesuada tortor ut nibh viverra eleifend.</p><p>Duis efficitur, arcu vitae viverra consectetur, nisi mi pharetra metus, vel egestas ex velit id leo. Curabitur non tortor nisi. Mauris ornare, tellus vel fermentum suscipit, ligula est eleifend dui, in elementum nunc risus in ipsum. Pellentesque finibus aliquet turpis sed scelerisque. Pellentesque gravida semper elit, ut consequat est mollis sit amet. Nulla facilisi.</p>";
for (var i = 0; i < 3; i++) {
  lipsum = lipsum + lipsum;
}

// Optional override to replace auto-generated contents page.
// If not required, just drop it from the 'var epub=' call below.
var generateContentsPage = function (links) {
  var contents = "<h1>Chapters</h1>";
  _.each(links, function (link) {
    // Omit all but the main pages.
    if (link.itemType === "main") {
      if (link.title === "More Books to Read") {
        contents += "<div> &nbsp;</div>";
      }
      contents += "<div><a href='" + link.link + "'>" + link.title + "</a></div>";
    }
  });
  return contents;
};

// Set up the EPUB basics.
var epub = nodepub.document(metadata, "test/test-cover.png", generateContentsPage);
epub.addCSS(`body { font-family:Verdana,Arial,Sans-Serif; font-size:11pt; }
#title,#title h1,#title h2,#title h3 { text-align:center; }
h1,h3,p { margin-bottom:1em; }
h2 { margin-bottom:2em; }
p { text-indent: 0; }
p+p { text-indent: 0.75em; }`);

// Add some front matter.
epub.addSection('Title Page', "<div id='title'><h1>[[TITLE]]</h1><h2>Book <strong>[[SEQUENCE]]</strong> of <em>[[SERIES]]</em></h2><h3>[[AUTHOR]]</h3><p> &nbsp;</p><p>&copy; [[COPYRIGHT]]</p></div>", true, true);
epub.addSection('Copyright', copyright, false, true);

// Add some content.
epub.addSection('Chapter 1', "<h1>One</h1>" + lipsum);
epub.addSection('Chapter 2', "<h1>Two</h1>" + lipsum);
epub.addSection('Chapter 2a', "<h1>Two (A)</h1><p><strong>This chapter does not appear in the contents.</strong></p>" + lipsum, true);
epub.addSection('Chapter 3', "<h1>Three</h1><p>Here is a sample list.</p><ul><li>Sample list item one.</li><li>Sample list item two.</li><li>Sample list item three.</li></ul>" + lipsum);
epub.addSection('More Books to Read', more);
epub.addSection('About the Author', about);

// Generate the result.
epub.writeEPUB(
  function (e) { console.log("Error:", e); },
  'example', 'example',
  function () { console.log("No errors. See the 'example' subfolder.") }
);

// Also write the structure both for debugging purposes and also to provide sample output in GitHub.
epub.writeFilesForEPUB('example/example-EPUB-files', (err) => {
  if (err) {
    console.log(err);
  }
});
